app/controllers/spree/adyen_redirect_controller.rb

```diff
@@ -67,6 +64,19 @@ def payment_method
       @payment_method ||= Gateway::AdyenHPP.last # find(params[:merchantReturnData])
     end
 
+    def current_order
+      @__adyen_current_order ||=
+        Spree::Order.incomplete.find_by!(current_order_params)
+    end
+
+    def current_order_params
+      { number: params[:merchantReference],
+        store_id: current_store.id,
+        user_id: try_spree_current_user.try(:id),
+        currency: current_currency
```

Comment:
I talked it over with @jhawthorn and in our opinion it would be best to only use the guest token to preform the lookup.

This token is exposed in the session as a signed cookie and is already publicly viewable in the request so there shouldn't be any associated security concerns.

This would look something like:
```ruby
cookies.permanent.signed[:guest_token] = params[:merchantReturnData] if params[:merchantReturnData]
@order = Order.incomplete.find_by!(guest_token: cookies.signed[:guest_token])
```
which is based on:  [OrdersCheckoutController.check_authorization](https://github.com/solidusio/solidus/blob/master/frontend/app/controllers/spree/orders_controller.rb#L85-L94) and [OrdersCheckoutController.edit](https://github.com/solidusio/solidus/blob/master/frontend/app/controllers/spree/orders_controller.rb#L37-L40)